Add gzip and zlib support for FixedLengthRecordReader

yongtang opened the pull request:
This fix adds gzip and zlib support for FixedLengthRecordReader,
as was discussed in #8865.

When FixedLengthRecordReader is used, it will check for
compression_type flag and use ZlibInputStream as needed.
The usage of InputBuffer in FixedLengthRecordReader has
also been changed to BufferedInputStream to match ZlibInputStream.

This fix fixes #8865.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

Mistobaan commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -28,45 +31,92 @@ namespace tensorflow {
-      *at_end = true;
-      return Status::OK();
+    if (file_pos_limit_ < 0) {
+      // In case of compressed file (non-seekable), attampt to read
```
*attempt

yongtang replied:
@Mistobaan Fixed. Thanks.

Mistobaan commented on tensorflow/core/ops/io_ops.cc:
```diff
@@ -442,6 +442,7 @@ REGISTER_OP("FixedLengthRecordReader")
     .Attr("footer_bytes: int = 0")
     .Attr("container: string = ''")
     .Attr("shared_name: string = ''")
+    .Attr("compression_type: string = ''")
```
I would suggest to use the word "encoding" instead of "compression_type"

yongtang replied:
Done.

yongtang commented:
@Mistobaan Thanks for the review. The PR has been updated. Please take a look.

drpngx commented:
@Mistobaan what do you think?

Mistobaan commented:
@drpngx looks fine to me

drpngx commented:
@saxenasaurabh please take a look

drpngx commented:
@saxenasaurabh LMK if you have cycles or we can find another reviewer.

saxenasaurabh commented on tensorflow/python/kernel_tests/reader_ops_test.py:
```diff
@@ -367,14 +367,43 @@ def _CreateFiles(self):
+          zf.write(cdata)
+    return filenames
+
+  def _TestOneEpoch(self, files, encoding=None):
```
Is the `files` param used? It gets overwritten on the next line.

yongtang replied:
Thanks @saxenasaurabh. That was a typo. The updated PR has fixed this issue by removing the  next line.

saxenasaurabh commented on tensorflow/python/kernel_tests/reader_ops_test.py:
```diff
@@ -390,6 +419,18 @@ def testOneEpoch(self):
+    files = self._CreateFiles()
+    self._TestOneEpoch(files)
+
+  def testGzipOneEpoch(self, encoding="GZIP"):
```
Change to
```
def testGzipOneEpoch(self):
  files = self._CreateGzipFiles()
  self._TestOneEpoch(files, encoding="GZIP")
```

saxenasaurabh reviewed (changes requested):
Thanks for doing this and sorry for the slow review.

yongtang commented:
Thanks @saxenasaurabh for the review. The reason that `footer_cache_` is needed is that `zlib/gzip` are essentially non-seekable. In other words, once a byte is read, we could not go backward. We can only start from the beginning again to find the previous location. In addition, `zlib/gzip` does not give us the uncompressed file size before hand. (we could certainly do a two-pass but that is likely inefficient if the file is large.)

In case of FixedlengthRecord, a footer might be possible. However, at the time we try to read the next bytes, we don't know if the next bytes belongs to a record, or a footer. Assume the record is `1` bytes and the footer is `2` bytes:
```
HRFF
```
When we read the first `R`, we don't know if it is record, or footer. We only knows it is a record when we read two `FF` + EOF. At this point, we knows that `FF` is the footer, thus `R` can only be record.

For that we maintains a `footer_cache_` so that we could find out the record `R` without seek back the file. (gzip/zlib is non-seekable and we could not find the length before hand).

I will update the PR to add a detailed explanation.

yongtang commented:
Also, it looks like the format of the `FixedLengthRecord` now accepts `hop_bytes` (added in another PR). This may complicate the processing for non-seekable files (like gzip/zlib).

I will take a closer look and see what we could do to handle `hop_bytes` as well.

yongtang committed:
Switch from InputBuffer to BufferedInputStream for FixedLengthRecordReader

This commit switch from InputBuffer to BufferedInputStream for
FixedLengthRecordReader, for the purpose of allowing compressed file as
FixedLengthRecordReader.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

vrv commented:
@yongtang let us know what you find.

yongtang committed:
Add gzip and zlib support for FixedLengthRecordReader
This fix adds gzip and zlib support for FixedLengthRecordReader,
as was discussed in 8856.
When FixedLengthRecordReader is used, it will check for
encoding flag and use ZlibInputStream as needed.
The usage of InputBuffer in FixedLengthRecordReader has
also been changed to BufferedInputStream to match ZlibInputStream.

This fix fixes 8856.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @saxenasaurabh @vrv. I managed to update the PR and the `hop_bytes_` should have been covered as well.

It use similar an approach of caching a buffer for `hop_bytes`.

Basically the purpose is to avoid the need to randomly seek to a position of the file stream. Now the file stream (possibly gzip or zlib) could be "read through" without moving backward. (gzip or zlib does not support moving backward as they are not "seekable").

The comment has been expanded to explain the logic but let me know if more explanation is needed.

yongtang commented:
I splited the PR into two commits. The first commit switches the InputBuffer to BufferedInputStream only (and drop the need of `Seek()` call). The second commit adds the gzip and zlib process. That may make it easy to follow.

vrv commented:
@yongtang it looks like somments from @saxenasaurabh haven't been addressed -- can you make sure or at least comment on them?  Thanks!

yongtang commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -28,45 +31,92 @@ namespace tensorflow {
+
+      // EOF reached
+      if (value->size() == 0) {
+        file_pos_limit_ = buffered_input_stream_->Tell() - footer_bytes_;
```
Thanks @saxenasaurabh. The updated PR includes additional comments for the explanation.

saxenasaurabh replied:
I still don't understand why you need to set file_pos_limit_ here?

yongtang commented:
@vrv @saxenasaurabh Thanks for the review. The review comments should have been covered. Please take a look.

drpngx commented:
Jenkins, test this please.

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -38,55 +42,178 @@ class FixedLengthRecordReader : public ReaderBase {
+      // footer_cache_: record_bytes_ --> end(record_bytes_ + footer_bytes_)
+      // Next time we read, we advance hop_bytes_ and append to the end, and
+      // repeat the same process.
+      if (0 < hop_bytes_ && hop_bytes_ < record_bytes_) {
```
Thanks for intricately handling hop_bytes and footer_bytes. It may be cleaner to combine the two caches into a `lookahead_cache_` which could be of size `record_bytes_ + footer_bytes`. You could use the same algorithm as you do now i.e. return record_bytes_ from lookahead_cache_ -> shift cache forward.

yongtang replied:
Thanks @saxenasaurabh. The PR has been updated with footer_cache and hop_cache combined. Also, now the `file_pos_limit_` seems to be not needed so I just removed the `file_pos_limit_` completely.

josh11b commented:
API review is fine with this change in principle, as long as it is backwards compatible (defaults for all new attrs, and those defaults give old behavior).

asimshankar commented on tensorflow/core/ops/io_ops.cc:
```diff
@@ -454,6 +455,7 @@ record_bytes: Number of bytes in the record.
 footer_bytes: Number of bytes in the footer, defaults to 0.
 hop_bytes: Number of bytes to hop before each read. Default of 0 means using
         record_bytes.
+encoding: The type of encoding for the file. Defaults to none.
```
Might be a good idea to elaborate on what valid values of this attribute can be.

yongtang replied:
@asimshankar Thanks. The PR has been updated with valid values (`GZIP` and `ZLIB`) covered.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented:
@yongtang any progress?

yongtang committed:
Remove the need to file the file size before hand for FixedLengthRecordReader

This commit updates the implementation of FixedLengthRecordReader so that
it will not try to find the file size before hand.

The purpose is to allow FixedLengthRecordReader to take a non-seekable file
as an input. In a non-seekable file (e.g., a compressed file), it may not
be possible to find the file size (e.g., in a uncompressed file length).

This commit is part of the effort to support gzip/zlib compression.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -38,55 +42,178 @@ class FixedLengthRecordReader : public ReaderBase {
+int bytes_to_read = record_bytes_;
+if (0 < hop_bytes_ && hop_bytes_ < record_bytes_) {
+bytes_to_read = hop_bytes_;
+}
```
Thanks @saxenasaurabh. The indent has been fixed.

yongtang commented:
Thanks all for the review and help and sorry for the delay. The PR has been updated. Please take a look and let me know if there are any issues.

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@saxenasaurabh can you take another look and approve?

martinwicke commented:
@saxenasaurabh Can you check this please? In particular make sure that all ops are backwards compatible.

I'm assuming this needs to be rebased.

martinwicke commented:
Jenkins, test this please.

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -29,73 +32,137 @@ class FixedLengthRecordReader : public ReaderBase {
+    // For example, assume record_bytes is 2 and hop_bytes is 1,
+    // For file "H123",
+    // We will process the data in the following way:
+    // 1. Read hope_bytes of 1 ("1") and save it in hop cache.
```
nit typo: hope_bytes -> hop_bytes

yongtang replied:
Done.

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -29,73 +32,137 @@ class FixedLengthRecordReader : public ReaderBase {
+    // In case hop_bytes_ is in between 0 and record_bytes_,
+    // we will need to hold a cache so that later on we could prefix the cache
+    // with the remaining data to read:
+    // For example, assume record_bytes is 2 and hop_bytes is 1,
```
and header_bytes is 1?

yongtang replied:
Thanks @saxenasaurabh. Done.

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -107,11 +174,15 @@ class FixedLengthRecordReader : public ReaderBase {
+  std::unique_ptr<RandomAccessFile> file_;
+  // must outlive buffered_input_stream_
+  std::unique_ptr<io::RandomAccessInputStream> file_stream_;
+  std::unique_ptr<io::InputStreamInterface> buffered_input_stream_;
```
May be call this `buffered_inputstream_` to be consistent with the file naming convention(buffered_inputstream.h etc).

yongtang replied:
Done.

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -29,73 +32,137 @@ class FixedLengthRecordReader : public ReaderBase {
+    if (0 < hop_bytes_ && hop_bytes_ < record_bytes_) {
+      bytes_to_read += record_bytes_ - hop_bytes_;
+    }
+    TF_RETURN_IF_ERROR(buffered_input_stream_->ReadNBytes(bytes_to_read, &lookahead_cache_));
```
What happens if the file is empty, i.e., has no records? ZlibInputStream would throw a OutOfRangeError here. May be defer filling cache till first call to ReadLocked?
It would be super awesome if you could add a test for this!

yongtang replied:
Thanks @saxenasaurabh. This part has also been addressed and additional test cases have been added.

drpngx commented:
It looks like the test fails because of an API change. @martinwicke could you confirm that it's OK?

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -29,73 +32,137 @@ class FixedLengthRecordReader : public ReaderBase {
-    if (hop_bytes_ > 0) {
-      input_buffer_->Seek(pos_before_read + hop_bytes_).IgnoreError();
+    if (hop_bytes_ > record_bytes_) {
+      buffered_input_stream_->SkipNBytes(hop_bytes_ - record_bytes_);
```
Here you would need to skip past `hop_bytes_ - record_bytes_` in the `lookahead_cache_` and not the `buffered_input_stream_` right? Then you would probably need to try to fill up the `lookahead_cache_` to `footer_bytes_`. If that is not possible you can set `*at_end = true`.
Could you add a test for this case as well please?

yongtang replied:
Thanks @saxenasaurabh. Yes the skip should happen at the beginning of the `lookahead_cache_` and `buffered_input_stream_` should read and append buffer to `lookahead_cache_`. I have updated the implementation and added the test cases for it.

saxenasaurabh reviewed (changes requested):
Really sorry for the delay in review.

yongtang committed:
Add additional test cases where num_records=0 and hop_bytes>record_bytes

This commit adds additional test cases where num_records=0 and hop_bytes>record_bytes

This commit also updated the API golden so that API compatibility test passes.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @saxenasaurabh for the review. The PR has been updated. Please take a look.

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -29,73 +32,155 @@ class FixedLengthRecordReader : public ReaderBase {
+    }
+
+    int bytes_to_read = record_bytes_;
+    if (0 < hop_bytes_) {
```
I know this has a lot of corner cases. Here's one suggestion for a simpler mental model:

In the constructor, set `hop_bytes_ = record_bytes_` if `hop_bytes_ == 0`. So now `hop_bytes_` is always > 0.

Remove the `buffered_inputstream_->ReadNBytes(footer_bytes_, &lookahead_cache_)` from `OnWorkStarted`.

In `ReadLocked`

1. Skip past `hop_bytes_` if not reading first record.
```cpp
if (record_number_ != 0) {
  if (hop_bytes_ <= lookahead_cache_.size()) {
    lookahead_cache_ = lookahead_cache_.substr(hop_bytes_);
  } else {
    int64 cache_size = lookahead_cache_.size();
    lookahead_cache_.clear();
    Status s = buffered_inputstream_->SkipNBytes(hop_bytes_-cache_size);
    if (!s.ok()) {
      if (!errors::IsOutOfRange(s)) {
        return s;
      }
      *at_end = true;
      return Status::OK();
    }
  }
}
```

2. Fill up `lookahead_cache_` to `record_bytes_ + footer_bytes_`.
```cpp
bytes_to_read = record_bytes_ + footer_bytes_ - lookahead_cache_.size();
Status s = buffered_inputstream_->ReadNBytes(bytes_to_read, value);
if (!s.ok()) {
  value.clear();
  if (!errors::IsOutOfRange(s)) {
    return s;
  }
  *at_end = true;
  return Status::OK();
}
lookahead_cache_.append(*value, 0, bytes_to_read);
value->clear();
```

3. Copy first `record_bytes_` from cache to `value`
```cpp
*value = lookahead_cache_.substr(0, record_bytes_);
```

yongtang replied:
Thanks @saxenasaurabh for the review. Yes that would be better to understand. The PR has been updated.

yongtang commented:
Thanks @saxenasaurabh for the review. The PR has been updated. Please take a look.

saxenasaurabh commented on tensorflow/core/kernels/fixed_length_record_reader_op.cc:
```diff
@@ -107,11 +142,15 @@ class FixedLengthRecordReader : public ReaderBase {
   const int64 record_bytes_;
   const int64 footer_bytes_;
   const int64 hop_bytes_;
+  string lookahead_cache_;
```
May be add a comment explaining
1. Why this is needed.
2. What the maximum size of this buffer will be.

yongtang replied:
Thanks @saxenasaurabh. The PR has been updated with comments added.

saxenasaurabh commented:
Jenkins, test this please.

yongtang commented:
Thanks @saxenasaurabh for the review. The PR has been updated. I also updated the api golden so that api compatibility could pass. Please take a look.

drpngx commented:
Jenkins, test this please.

josh11b commented on tensorflow/core/ops/io_ops.cc:
```diff
@@ -448,6 +448,7 @@ REGISTER_OP("FixedLengthRecordReader")
     .Attr("record_bytes: int")
     .Attr("footer_bytes: int = 0")
     .Attr("hop_bytes: int = 0")
+    .Attr("encoding: string = ''")
```
There should be no new uses of the V1 op, so there should be no reason to change it.

yongtang replied:
Thanks @josh11b, done.

josh11b commented on tensorflow/core/ops/io_ops.cc:
```diff
@@ -473,6 +476,7 @@ REGISTER_OP("FixedLengthRecordReaderV2")
     .Attr("record_bytes: int")
     .Attr("footer_bytes: int = 0")
     .Attr("hop_bytes: int = 0")
+    .Attr("encoding: string = ''")
```
Please move this attr to the end, to match the order in the Python API.

yongtang replied:
Thanks @josh11b, done.

josh11b commented on tensorflow/python/ops/io_ops.py:
```diff
@@ -396,6 +396,7 @@ def __init__(self,
                header_bytes=None,
                footer_bytes=None,
                hop_bytes=None,
+               encoding=None,
```
For backwards-compatibility reasons, please put this new parameter last.

yongtang replied:
Done.

josh11b reviewed (commented):
(Comments from API review)

drpngx reviewed (changes requested):
Marking changes request to avoid accidentally merging.

yongtang commented:
Thanks @josh11b @drpngx for the review. The PR has been updated. Please take a look.

drpngx commented:
Jenkins, test this please.

drpngx reviewed (approved):
Verified that the latest version addressed @josh11b comments.
